include/openthread/platform/mdns_socket.h

```diff
@@ -0,0 +1,172 @@
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for mDNS socket.
+ *
+ */
+
+#ifndef OPENTHREAD_PLATFORM_MULTICAST_DNS_SOCKET_H_
+#define OPENTHREAD_PLATFORM_MULTICAST_DNS_SOCKET_H_
+
+#include <stdint.h>
+
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/message.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup plat-mdns
+ *
+ * @brief
+ *   This module defines platform APIs for Multicast DNS (mDNS).
+ *
+ * @{
+ *
+ */
+
+/**
+ * Represents a socket address info.
+ *
+ */
+typedef struct otPlatMdnsAddressInfo
+{
+    otIp6Address mAddress;      ///< IP address. IPv4-mapped IPv6 format should be used to represent IPv4 address.
+    uint16_t     mPort;         ///< Port number.
+    uint32_t     mInfraIfIndex; ///< Interface index.
+} otPlatMdnsAddressInfo;
+
+/**
+ * Enables or disables platform mDNS operation.
+ *
+ * Once enabled, platform SHOULD start listening for mDNS messages sent to UDP destination port 5353 (well-known port
+ * assigned to mDNS).
+ *
+ * Platform MUST listen for multicast messages sent to UDP destination port 5353 at the mDNS link-local multicast
+ * address `224.0.0.251` and/or its IPv6 equivalent `ff02::fb`.
+ *
+ * Platform SHOULD also listen for any unicast messages sent to UDP destination port 5353. If this is not possible,
+ * then OpenThread mDNS module can be configured so to not use any "QU" questions requesting unicast response.
+ *
+ * All received messages MUST be reported back using `otPlatMdnsHandleReceive()` callback.
+ *
+ * @param[in] aInstance  The OpernThread instance.
+ * @param[in] aEnable    Indicate whether to enable or disable.
+ *
+ */
+void otPlatMdnsSetEnabled(otInstance *aInstance, bool aEnable);
+
+/**
```

Comment:
This API doesn't specify on which interface to send the multicast packet

Reply:
This API requires the platform to multicast the message over all network interfaces within the mDNS multicast domain. The platform is responsible to manage the set of relevant interfaces.

My idea is that if a future use case arises where we need to multicast on a specific interface, we introduce a new separate API for that purpose. This would make this behavior explicit and clear.

Reply:
Border routing currently supports only a single link, if a SRP service is advertised on all mdns-capable links, the service will be discoverable on a non-AIL link but **not reachable**. I think this is worse than making SRP service discoverable on only AIL link. So for SRP, it's better to specify the desired the advertising interface. Thoughts?

Reply:
I agree. Maybe to phrase it better we should say "this API should multicast on all network interfaces where device is acting as a Border Router".  Today that is limited to one interface. I will add an parameter to specify the interface index in the next push to this API.

Reply:
@wgtdkp the new pushed commit adds the interface index to this platform API and other APIs.

Reply:
Thanks abtink! resolving this comment.